WIP: Port app to be deployable to Heroku or Cloud Foundry

konklone opened the pull request:
This PR tracks progress getting the app prepared and deployed on **Cloud Foundry** (hereafter referred to as **CF**). Since CF uses Heroku buildpacks and philosophy, the work as of PR-file-time is to make the app run on Heroku, [which it does](https://foia-test-18f.herokuapp.com).

I've updated the README, removed some deprecated `--use-mirrors` flags from `.travis.yml`, and switched out `gunicorn` for `waitress` in the project's `requirements.txt`.

Some outstanding work:

* Figure out how we're going to retool our data loading process to run entirely on a node with an ephemeral, non-guaranteed disk. Cloning a git repo ain't going to work. One approach: host the data on GitHub Pages and have the loading script make a GET request to fetch each YAML file on demand. Another: back up the YAML files to S3, automatically if possible, and download them from there instead. Created #511.
* Continue with any remaining Cloud Foundry specific work. (One known tweak: `$PORT` -> `$VCAP_APP_PORT`.)
* Investigate RDS instead of a postgres service directly.

There's some documentation in `cloudfoundry.md` that I'm keeping along the way, it's not very formal.

Fixes #501.
Fixes #491

khandelwal commented:
Why are we using waitress instead of gunicorn?

konklone commented:
In part because it hews to the [Django/CF guide that @dlapiduz started](https://github.com/18F/cloud-foundry-notes/blob/master/deploying_django.md), and in part because [gunicorn expects to be behind a buffering reverse proxy](http://blog.etianen.com/blog/2014/01/19/gunicorn-heroku-django/) and is a poor candidate for Heroku. I'm assuming that Cloud Foundry has a similar design to Heroku in this regard, though I'm not 100% certain.

khandelwal commented:
Ah, thank you. I understand that much better now.

khandelwal commented on requirements.txt:
```diff
@@ -10,8 +10,11 @@ restless==2.0.1
-dj-static
+# deployed/hosted with waitress
+waitress==0.8.9
+dj-static==0.0.6
```
Why do we need this? What does it do?

konklone replied:
dj-static was already a requirement. It handles static files in a WSGI context. Beyond that -- I'm just following instructions, and don't know much more.

khandelwal commented on requirements.txt:
```diff
@@ -10,8 +10,11 @@ restless==2.0.1
+dj-static==0.0.6
+
+# use database strings instead of dicts
+dj-database-url==0.3.0
```
I'm assuming this is so that we can store the database connection information in a single environment variable?

konklone replied:
Correct, a common practice in Heroku-ish land.

khandelwal commented on runtime.txt:
```diff
@@ -0,0 +1 @@
+python-3.4.2
```
What's this file for?

konklone replied:
Tells Heroku and Cloud Foundry to use Python 3.4.2 as the language runtime. This is separate from specifying that it's a "Python app", which it infers from the presence of `requirements.txt`.

khandelwal commented:
Other than a few questions I'd like answered for my own edification - this looks good.

khandelwal commented:
Are we ready to co-ordinate to merge this in? @ramirezg and @konklone Can you work on this together?

konklone commented:
@ramirezg and I can work on merging this in. Bear in mind, progress on this stalled since my last commit here, as 18F's internal Cloud Foundry instance was migrated, updated, and fixed. @dlapiduz said that that's settled now, and so -- _after_ merging this in and seeing it running on our AWS server and at the Heroku URL with the same code, we can do a separate PR with CloudFoundry-specific changes.
